[rocrand] Add support for gfx803

tpkessler opened the pull request:
For unknown reason, `cmake` overwrites the list `AMDGPU_TARGETS`, thereby removing `gfx803`. Manually specifying `AMDGPU_TARGETS` fixes the problem.
I've created this PR because of https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/issues/1106#issuecomment-689536115

oleid commented:
I still have no clue why it overwrites this, still debugging.

tpkessler commented:
I found it! `AMDGPU_TARGETS` is defined with the `CACHE` attribute so I looked for included cmake files that also define this variable. Indeed:

```bash
/opt/rocm/hip/lib/cmake/hip/hip-config.cmake:106:  set(AMDGPU_TARGETS "gfx900;gfx906;gfx908" CACHE STRING "AMD GPU targets to compile for")
```
which is included by `cmake/VeriftyCompiler.cmake`. The file is part of `hip-rocclr`.

oleid commented:
I manually added it for testing, like you did. But I think I found the cause. The variable is a cache variable and already set before it in the top level `CMakeLists.txt`. Thus, it is not overwritten (unless we force it).

It would seem that the variable is already predefined in

`/opt/rocm/hip/lib/cmake/hip/hip-config.cmake`.

Thus, it would probably make sense _not_ to add this patch but to change the default in `hip-config.cmake`

oleid commented:
A, you were 5 seconds faster :smile:

tpkessler commented:
> Thus, it would probably make sense not to add this patch but to change the default in hip-config.cmake

Absolutely. I'll try to patch this for HIP.

oleid commented:
I find it puzzling that this GPU isn't included in the defaults:

```
./hip-config.cmake.in:  set(AMDGPU_TARGETS "gfx900;gfx906;gfx908" CACHE STRING "AMD GPU targets to compile for")
```

Do you think it would make sense to include Navi (`gfx1010`) as well? Not sure how complete they are. And I don't have such a GPU.

oleid commented:
According to phoronix (https://www.phoronix.com/scan.php?page=news_item&px=GFX1011-GFX1012-AMDGPU-LLVM) big navy could be supported as well (gfx1011 and 1012). At least xuhuisheng  has gfx1011 in the list of their files.

tpkessler commented:
> Do you think it would make sense to include Navi

It certainly does, but I can't test it (I'm on a Vega 56) and I'm not aware of a regular contributor to this repo that owns one. Navi support is not fully fleshed out from AMD's side, but we should include it. If it works, then we're fine. Otherwise people should file upstream issue and pressure AMD to finally fully support it.

tpkessler commented:
> AMDGPU-LLVM

Yes, `clang` from `llvm-amdgpu` is able to emit bite code for Navi. From the phoronix forum I also know that OpenCL is supposed to work on Navi. The only issue might be that not all libraries (rocblas, rocrand, rocfft) include the assembly code for Navi.

tpkessler commented:
> Do you think it would make sense to include Navi

I've added it to the list but then building `rocrand` fails because of unsupported asm instructions (as expected). That's why I've removed it again.

u9rzm commented:
Guys! Help me please. No matter how much I understand, I can not understand how you solved this problem. Rebuilt through HIP? or just change hip-config.cmake to `('gfx803').` If the first, please tell me how to assemble it? Or can there be a package directly?
I've been crying over this for a week((

Error: 
`2021-01-16 01:36:57.395956: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libamdhip64.so
/src/external/hip-on-vdi/rocclr/hip_code_object.cpp:120: guarantee(false && "hipErrorNoBinaryForGpu: Coudn't find binary for current devices!")
Aborted (core dumped)`

my list

`root@u9-rzm:/tf_docker_share# find /opt/rocm/rocblas/ -type f
/opt/rocm/rocblas/include/rocblas_module.f90
/opt/rocm/rocblas/include/internal/rocblas_device_malloc.hpp
/opt/rocm/rocblas/include/internal/rocblas-functions.h
/opt/rocm/rocblas/include/internal/rocblas-types.h
/opt/rocm/rocblas/include/internal/rocblas_bfloat16.h
/opt/rocm/rocblas/include/internal/rocblas-complex-types.h
/opt/rocm/rocblas/include/internal/rocblas-exported-proto.hpp
/opt/rocm/rocblas/include/internal/rocblas-auxiliary.h
/opt/rocm/rocblas/include/internal/rocblas-version.h
/opt/rocm/rocblas/include/internal/rocblas-export.h
/opt/rocm/rocblas/include/rocblas.h
/opt/rocm/rocblas/lib/library/Kernels.so-000-gfx908.hsaco
/opt/rocm/rocblas/lib/library/Kernels.so-000-gfx1011.hsaco
/opt/rocm/rocblas/lib/library/Kernels.so-000-gfx906.hsaco
/opt/rocm/rocblas/lib/library/Kernels.so-000-gfx900.hsaco
/opt/rocm/rocblas/lib/library/Kernels.so-000-gfx803.hsaco
/opt/rocm/rocblas/lib/library/TensileLibrary_gfx900.co
/opt/rocm/rocblas/lib/library/Kernels.so-000-gfx1010.hsaco
/opt/rocm/rocblas/lib/library/TensileLibrary_gfx906.co
/opt/rocm/rocblas/lib/library/TensileLibrary_gfx908.co
/opt/rocm/rocblas/lib/library/TensileLibrary_gfx803.co
/opt/rocm/rocblas/lib/library/TensileLibrary.dat
/opt/rocm/rocblas/lib/cmake/rocblas/rocblas-config.cmake
/opt/rocm/rocblas/lib/cmake/rocblas/rocblas-targets.cmake
/opt/rocm/rocblas/lib/cmake/rocblas/rocblas-config-version.cmake
/opt/rocm/rocblas/lib/cmake/rocblas/rocblas-targets-release.cmake
/opt/rocm/rocblas/lib/librocblas.so.0.1.40000`

undying commented:
@u9rzm some time ago I've spent weeks to build this and made a docker file to remember the process.
You may try to use it to build rocm https://github.com/undying/ROCm-Docker/blob/master/Dockerfile.pytorch
